search-improvements.rst

```diff
@@ -0,0 +1,143 @@
+   :author: Safwan
+   :location: DHA
+
+Improved Search
+==================================
+Have you ever used the search functionality inside documentations hosted by Read The Docs?
+Mostly yes. Search can always be improved and in the era of Search Engines, the
+documentation search also plays a big role for the users. We have realized the importance,
+and got our GSoC student interested to take the challenge. The main goal was to refactor the
+search code together with upgrading the backend Search engine, as well as adding more features
+to the search functionality like exact match search, case insensitive search,
+search as you type, suggestions etc.
+
+Background
+^^^^^^^^^^
+Search is a vital part of any documentation hosting platform as people can get the
+information they need. As a documentation hosting platform, same rule applies for
+Read The Docs. Because of having a small core team, the search functionality
+of Read The Docs has lagged behind for quite a while now. Initially the search code
+was voluntarily contributed by `Rob Hudson`_,  `back in 2013`_ and then improved by other
+contributors. The time span and continuous technology growing made the search
+module got older and it had become outdated. Moreover The search infrastructure was already
+vulnerable as we were using `Elasticsearch 1.3.x`_ which was already reached its
+End of Life in 2016. Therefore, the upgradation of search infrastructure was badly needed.
+
+Google Summer of Code
+~~~~~~~~~~~~~~~~~~~~~
+Google Summer of Code is a global program where students work with an open source organization
+on a 3 month programming project. The core team of Read The Docs proposed some `Project Ideas`_,
+one of them was **Refactor & improve our search code**. Our contributor (me) `Safwan Rahman`_,
+an undergraduate university student and open source contributor from Bangladesh was keen
+to get his hand dirty with Elasticsearch_ and grasped the opportunity to do so by applying
+for this project and got accepted.
+
+Safwan_ has worked full time for last 3 months to upgrade the whole codebase
+to compatable with `Elasticsearch 6.x`_ and also implemented various features
+like `Exact Matching Search`_, `Case Insensitive Search`_, `Improved Result Order`_,
+`Zero Downtime Indexing`_ and many more. Together with this, he is willing to implement
+`Search as You Type and Autocomplete`_, `Code Search`_ and many more.  All of his search related work can be
+looked in the `Search Project Board`_.
+
+
+Sphinx/MkDocs Built in Search vs Read The Docs Search
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Sphinx/MkDocs already have built in search functionality. But the features are very limited and heavy.
+At Read The Docs, we have felt the limitations and therefore we index the documentations in our
+Elasticsearch_ index so that we can provide better results to the users. Currently, we do not index
+MkDocs documents to Elasticsearch_, but `any kind of help is welcome`_.
```

Comment:
I think we can expand a little bit on this section. Just saying it is heavy doesn't say much. Here's a few brief ideas worth mentioning:

- searching across different projects
- more advanced query syntax